beartype_test/a00_unit/a00_util/text/test_utiltextmunge.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+# --------------------( LICENSE                           )--------------------
+# Copyright 2014-2021 by Cecil Curry.
+# See "LICENSE" for further details.
+
+"""
+**Beartype replace utility unit tests.**
+
+This submodule unit tests the public API of the private
+:mod:`beartype._util.text.utiltextmunge.py` submodule.
+"""
+
+
+# ....................{ IMPORTS                           }....................
+# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+# WARNING: To raise human-readable test errors, avoid importing from
+# package-specific submodules at module scope.
+# !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+# from pytest import raises
+
+# ....................{ TESTS                             }....................
+def test_uppercase_char_first():
+    # Defer heavyweight imports.
+    from pytest import raises
+
+    from beartype._util.text.utiltextmunge import uppercase_char_first
+
+    # Not a string
+    with raises(AssertionError):
+        uppercase_char_first(7)
+
+    assert uppercase_char_first(text='beArTyPe') == "BeArTyPe"
+    assert uppercase_char_first(text="<bear>") == "<bear>"
+    assert uppercase_char_first(text="") == ""
+
+
+def test_number_lines():
+    # Defer heavyweight imports.
+    from pytest import raises
+
+    from beartype._util.text.utiltextmunge import number_lines
+
+    with raises(AssertionError):
+        number_lines(7)
+
+    # No lines to split
+    assert number_lines(text="") == ""
+
+    # This is in essence "x == x" - I don't know if this adds
+    # anything substantive...
+    text = "bears, beats, battlestar galactica\n" * 20
+    expected_result = "\n".join(
+        '(line {:0>4d}) {}'.format(line_no, line_text)
+        for line_no, line_text in enumerate(text.splitlines(), start=1))
+
+    assert number_lines(text) == expected_result
+
+
+def test_replace_str_substrs():
+    # Defer heavyweight imports.
+    from pytest import raises
+
+    from beartype._util.text.utiltextmunge import replace_str_substrs
+    from beartype.roar import _BeartypeUtilTextException
+
+    with raises(AssertionError):
+        replace_str_substrs(text=7, old='Oh No!', new='A non-str value')
+        replace_str_substrs(text='Oh No!', old=7.0, new='is a non-str value')
+        replace_str_substrs(text='Oh No!', old='A non-str value of ', new=7)
```

Comment:
> 
> 
> **Bwahah!** A growth opportunity emerges from the murk of mundanity like Godzilla preparing to subdue downtown Tokyo _yet again._

Excellent, let us step face first into a potentially embarrassing situation in the name of learning! :godmode:  

> Depressingly, the above `with ...:` block doesn't behave like we all wished it would. It _does_ correctly assert the first `replace_str_substrs()` call to raise the expected exception. That's good. But it also immediately exits that block after doing so, ensuring that the following two `replace_str_substrs()` calls are _never_ actually called. That's bad.
>....
>... a `wait ...:` block _immediately_ halts on the first uncaught exception, which means you can't raise multiple uncaught exceptions from the same `wait ...:` block, which means the only way to catch multiple uncaught exceptions with `wait ...:` blocks is to **wrap each statement expected to raise an uncaught exception in its own block:** e.g.,
> 
> ```
>     with raises(AssertionError):
>         replace_str_substrs(text=7, old='Oh No!', new='A non-str value')
> 
>     with raises(AssertionError):
>         replace_str_substrs(text='Oh No!', old=7.0, new='is a non-str value')
> 
>     with raises(AssertionError):
>         replace_str_substrs(text='Oh No!', old='A non-str value of ', new=7)
> ```
> 

How unorthodox would it be if instead of wrapping each individual statement, we wrapped the `wait...:` in a for loop? Specifically:

```
    raise_assertion_err = [
        "replace_str_substrs(text=7, old='Oh No!', new='A non-str value')",
        "replace_str_substrs(text='Oh No!', old=7.0, new='is a non-str value')",
        "replace_str_substrs(text='Oh No!', old='A non-str value of ', new=7)",
    ]
    for test in raise_assertion_err:
        with raises(AssertionError):
            eval(test)
```
If I'm understanding the `wait...:` block, this would solve the problem as contextmanager's `__exit__` should halt right into the next thing to test. If this IS the case, then my solution would keep the verbosity to a minimum while, hopefully, being straightforward... That said, I don't know if this is 'good form' - discouraged because of some inherent flaw.

Reply:
> How unorthodox would it be if instead...

**Heh.** The answer to any question prefixed by that sentence fragment is guaranteed to be "Very. Very unorthodox." This is no exception. Theoretically, that works. Pragmatically, that still works. Stylistically, that opens the perfidious doors to [Kralizec](https://dune.fandom.com/wiki/Kralizec/XD). <sup>*Curse you, Leo Atredes!*</sup>

Why? Please chant after me, castrati child choir: [`eval()` considered harmful.](https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html) `eval()` is slow (which doesn't *really* matter in tests), `eval()` is fragile (which *really* matters in tests), `eval()` is insecure (which *really* matters everywhere), and `eval()` obstructs introspection, type-checking, and AST manipulation (which sometimes *really* matters). We don't want our Arch Nemesis Who Shall Remain Nameless to raise a [CVE](https://cve.mitre.org) against `beartype`, so we don't `eval()`.

Of course, `beartype` already calls `eval()` once to resolve [PEP 563-deferred annotations](https://www.python.org/dev/peps/pep-0563/) – but **only because we have to** and **only because that's the official solution documented in PEP 563.** The general rule of thumb is that one should *never* `eval()` or `exec()` unless one has no other recourse, which one does here, so let's just do that other recourse.

Yes, boilerplate sucketh the sacrilegious knob of Satan, which is bad. But everyone will justifiably mock us until we feel bad about ourselves if we abuse `eval()` and `exec()` in tests without just cause, which is even worse. *Satan it is!*